File: 2019/rust/src/d3/p2.rs
use std::fs;
use std::collections::HashMap;
use std::ops::Add;

#[derive(Debug, Hash, Copy, Clone)]
struct Position {
    x: i32,
    y: i32
}

impl Add for Position {
    type Output = Position;

    fn add(self, rhs: Position) -> Position {
        Position{
            x: self.x + rhs.x,
            y: self.y + rhs.y,
        }
    }
}

impl PartialEq for Position {
    fn eq(&self, other: &Self) -> bool {
        self.x == other.x && self.y == other.y
    }
}

impl Eq for Position {}

pub fn main() {
    let fname = "src/d3/input.txt";
    //let fname = "src/d3/test_one.txt"; // 159
    let data: Vec<Vec<(char, i32)>> = fs::read_to_string(fname)
        .expect("read failed")
        .trim().lines() // &str to &[str]
        .map(|line| { // line: &str
            line.trim().split(",") // &[str]
            .map(|t| { // t: &str
                t.chars()
                .collect::<Vec<char>>() // &str to iter[Vec<char>]
            })
            .map(|t| { // t: Vec<char>
                (
                    t[0], // char
                    t[1..].iter()
                        .collect::<String>() // iter[char] to String
                        .parse().unwrap() // String to i32
                )
            }).collect() // Vec<(char, i32)>
        }).collect();

    // Track wires
    // Position to steps for first wire
    let mut hm: HashMap<Position, i32> = HashMap::new();

    // Track intersections
    // For each intersection, store the
    // combined number of steps
    let mut inter: Vec<(Position, i32)> = Vec::new();

    let get_delta = |d| {
        match d {
            'U' => Position{x: 1, y: 0},
            'R' => Position{x: 0, y: 1},
            'D' => Position{x: -1, y: 0},
            'L' => Position{x: 0, y: -1},
            _ => Position{x: 0, y: 0},
        }
    };

    let mut pos = Position{x: 0, y: 0};
    let mut steps = 0;
    for (dir, qty) in &data[0] {
        let d = get_delta(*dir);
        for _ in 0..*qty {
            steps += 1;
            pos = pos + d;
            hm.insert(pos, steps);
        }
    }

    let mut pos = Position{x: 0, y: 0};
    let mut steps = 0;
    for (dir, qty) in &data[1] {
        let d = get_delta(*dir);
        for _ in 0..*qty {
            steps += 1;
            pos = pos + d;
            match hm.get(&pos) {
                Some(w1_steps) => inter.push((pos, w1_steps + steps)),
                None => ()
            }
        }
    }

    let ans = inter.iter().map(|p| p.1).min();
    println!("{:?}", ans.unwrap());
}

File: 2019/Python/D3/p2.py
lines = [x[:-1] for x in open("input.txt").readlines()]
w1, w2 = lines[0], lines[1]

w1 = w1.split(",")
w2 = w2.split(",")

def hash(coord):
    return str(coord[0]) + ":" + str(coord[1])

def mapWire(wire):
    coord = [0, 0]
    map = {}
    steps = 0

    for part in wire:
        dir = part[0]
        dist = int(part[1:])

        if dir == "U":
            for _ in range(dist):
                map[hash(coord)] = steps
                coord[1] += 1
                steps += 1

        elif dir == "R":
            for _ in range(dist):
                map[hash(coord)] = steps
                coord[0] += 1
                steps += 1

        elif dir == "D":
            for _ in range(dist):
                map[hash(coord)] = steps
                coord[1] -= 1
                steps += 1

        else: # dir == "L"
            for _ in range(dist):
                map[hash(coord)] = steps
                coord[0] -= 1
                steps += 1

    return map

def sol():
    m1 = mapWire(w1)

    hit = []
    coord = [0, 0]
    steps = 0

    for part in w2:
        dir = part[0]
        dist = int(part[1:])

        if dir == "U":
            for _ in range(dist):
                if hash(coord) in m1:
                    hit.append([coord[0], coord[1], steps, m1[hash(coord)]])
                steps += 1
                coord[1] += 1

        elif dir == "R":
            for _ in range(dist):
                if hash(coord) in m1:
                    hit.append([coord[0], coord[1], steps, m1[hash(coord)]])
                steps += 1
                coord[0] += 1

        elif dir == "D":
            for _ in range(dist):
                if hash(coord) in m1:
                    hit.append([coord[0], coord[1], steps, m1[hash(coord)]])
                steps += 1
                coord[1] -= 1

        else: # dir == "L"
            for _ in range(dist):
                if hash(coord) in m1:
                    hit.append([coord[0], coord[1], steps, m1[hash(coord)]])
                steps += 1
                coord[0] -= 1

    # Find min steps
    best = float("inf")
    best_index = 0
    for i in range(len(hit)):
        h = hit[i]
        dist = abs(h[2]) + abs(h[3])

        if h[0] == 0 and h[1] == 0:
            continue

        if dist < best:
            best_index = i
            best = dist

    return best

ans = sol()
print(ans)
